hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/ITestRestrictedReadAccess.java

```diff
@@ -0,0 +1,459 @@
+    return newAssumedRoleConfig(getContract().getConf(), roleARN);
+  }
+
+
+  @Test
+  public void testNoReadAccess() throws Throwable {
+    describe("Test failure handling of if the client doesn't"
+        + " have read access under a path");
+    S3AFileSystem realFS = getFileSystem();
+
+    // avoiding the parameterization to steer clear of accidentally creating
+    // patterns
+    Path basePath = path("testNoReadAccess-" + name);
+    Path noReadDir = new Path(basePath, "noReadDir");
+    Path emptyDir = new Path(noReadDir, "emptyDir");
+    realFS.mkdirs(emptyDir);
+    Path emptyFile = new Path(noReadDir, "emptyFile.txt");
+    touch(realFS, emptyFile);
+    Path subDir = new Path(noReadDir, "subDir");
+
+    Path subdirFile = new Path(subDir, "subdirFile.txt");
+    byte[] data = "hello".getBytes(Charset.forName("UTF-8"));
+    createFile(realFS, subdirFile, true, data);
+    Path subDir2 = new Path(noReadDir, "subDir2");
+    Path subdir2File1 = new Path(subDir2, "subdir2File1.txt");
+    Path subdir2File2 = new Path(subDir2, "subdir2File2.docx");
+    createFile(realFS, subdir2File1, true, data);
+    createFile(realFS, subdir2File2, true, data);
+
+    // create a role filesystem which does not have read access under a path
+    Configuration roleConfig = createAssumedRoleConfig();
+    bindRolePolicyStatements(roleConfig,
+        STATEMENT_S3GUARD_CLIENT,
+        STATEMENT_ALLOW_SSE_KMS_RW,
+        statement(true, S3_ALL_BUCKETS, S3_ALL_OPERATIONS),
+        new Statement(Effects.Deny)
+            .addActions(S3_PATH_RW_OPERATIONS)
+            .addResources(directory(noReadDir)));
+    roleFS = (S3AFileSystem) basePath.getFileSystem(roleConfig);
+
+    // this is a LIST call; there's no marker.
+    roleFS.listStatus(basePath);
+
+    // this is HEAD + "/" on S3; get on S3Guard auth
+    roleFS.listStatus(emptyDir);
+
+    lsR(roleFS, noReadDir, true);
+
+    roleFS.getFileStatus(noReadDir);
+    roleFS.getFileStatus(emptyDir);
+    if (authMode) {
+      // auth mode doesn't check S3, so no failure
+      roleFS.getFileStatus(subdirFile);
+    } else {
+      intercept(AccessDeniedException.class, () ->
+          roleFS.getFileStatus(subdirFile));
+    }
+    intercept(AccessDeniedException.class, () ->
+        ContractTestUtils.readUTF8(roleFS, subdirFile, data.length));
+    if (authMode) {
+      // auth mode doesn't just not check the store,
+      // because it knows the file length is zero,
+      // it returns -1 without even opening the file.
+      try (FSDataInputStream is = roleFS.open(emptyFile)) {
+        Assertions.assertThat(is.read())
+            .describedAs("read of empty file")
+            .isEqualTo(-1);
+      }
+      roleFS.getFileStatus(subdirFile);
+    } else {
+      // non-auth mode, it fails at some point in the opening process.
+      intercept(AccessDeniedException.class, () ->
+          ContractTestUtils.readUTF8(roleFS, emptyFile, 0));
+    }
+
+    // Fun with Glob
```

Comment:
I would move this to another test. If something fails somebody (me) will need to debug and it could be harder for a test this long.